src/main.rs

```diff
@@ -1,10 +1,42 @@
-use midir::os::unix::VirtualOutput;
-use midir::MidiOutput;
+use midir::os::unix::{VirtualInput, VirtualOutput};
+use midir::{MidiInput, MidiOutput};
+use std::sync::{Arc, Mutex};
+use wmidi::MidiMessage;
+
+use line_runner::{LineLauncher, Message, MidiClockTracker};
 
 fn main() {
     let midi_out = MidiOutput::new("Line runner").unwrap();
 
-    let /*mut*/ _conn_out = midi_out.create_virtual("Line runner").unwrap();
+    let conn_out = midi_out.create_virtual("Line runner").unwrap();
+
+    let midi_in = MidiInput::new("Line runner").unwrap();
+
+    let (midi_clock_tracker, beat_message_receiver) = MidiClockTracker::new();
+    let midi_clock_tracker = Arc::new(Mutex::new(midi_clock_tracker));
```

Comment:
I don't know if this is a reasonable usage of `Arc<Mutex<_>>` or not - where I ran into problems was in trying to reference the MIDI clock tracker from inside the closure passed to `.create_virtual()` below, Rust thought the closure might outlive `midi_clock_tracker`

Reply:
It seems to be avoidable (although I don't have the best grasp of what exactly my solution is doing):

```diff
From c5717db2b3d6f57c428bbb47d4662f92a75628db Mon Sep 17 00:00:00 2001
From: Peter Stuart <peter@peterstuart.org>
Date: Fri, 18 Jun 2021 10:22:58 -0400
Subject: [PATCH] Remove Arc/Mutex.

---
 src/main.rs | 11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/src/main.rs b/src/main.rs
index aec6c61..4a51972 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -1,6 +1,5 @@
 use midir::os::unix::{VirtualInput, VirtualOutput};
 use midir::{MidiInput, MidiOutput};
-use std::sync::{Arc, Mutex};
 use wmidi::MidiMessage;
 
 use line_runner::{LineLauncher, Message, MidiClockTracker};
@@ -12,16 +11,14 @@ fn main() {
 
     let midi_in = MidiInput::new("Line runner").unwrap();
 
-    let (midi_clock_tracker, beat_message_receiver) = MidiClockTracker::new();
-    let midi_clock_tracker = Arc::new(Mutex::new(midi_clock_tracker));
-    let midi_clock_tracker_clone = midi_clock_tracker.clone();
+    let (mut midi_clock_tracker, beat_message_receiver) = MidiClockTracker::new();
 
     let _conn_in = midi_in
         .create_virtual(
             "Line runner",
             move |timestamp, bytes, _| {
                 if let Some(message) = Message::from(timestamp, bytes).unwrap() {
-                    handle_message(message, &midi_clock_tracker_clone);
+                    handle_message(message, &mut midi_clock_tracker);
                 }
             },
             (),
@@ -32,10 +29,10 @@ fn main() {
     line_launcher.listen();
 }
 
-fn handle_message(message: Message, midi_clock_tracker: &Arc<Mutex<MidiClockTracker>>) -> () {
+fn handle_message(message: Message, midi_clock_tracker: &mut MidiClockTracker) -> () {
     match message.message {
         MidiMessage::TimingClock => {
-            midi_clock_tracker.lock().unwrap().tick();
+            midi_clock_tracker.tick();
         }
         _ => {}
     }
-- 
2.32.0
```

I *think* what is happening here is that `midi_clock_tracker` is being moved into the closure, and then the body of the closure uses a mutable reference to it. `midi_clock_tracker` is not available after the definition of the closure, because of having been moved.

I'd be curious to see what you tried. I wasn't actually expecting this to work!

Reply:
I think it's kind of cool that you can include a patch in comments like this and get it highlighted all nicely.

Reply:
Ok ya now that I have Clippy turned on it led me to this solution by alerting me to the fact that the `.clone()` (of the `Arc`) was "redundant" - basically when I was writing it, I thought I needed two separate references to the MIDI clock tracker. But I think I ended up avoiding that by using a channel for having it communicate beats back to the line launcher

> I think it's kind of cool that you can include a patch in comments like this and get it highlighted all nicely.

Ya that works surprisingly nicely as a "diff within a diff"